### PR TITLE
Add zoom to image annotation and check jpeg file quality before save [SCI-3300]

### PR DESCRIPTION
**Pull request by aignatov-bio**
Jira ticket: [SCI-3300](https://biosistemika.atlassian.net/browse/SCI-3300)

### What was done
Adding ability to zoom image with mouse wheel (not perfect).
Check image quality for jpeg before update.

### CODE REVIEW

File: app/assets/javascripts/sitewide/file_preview.js [hunk 1]
```diff
@@ -278,6 +276,37 @@ var FilePreviewModal = (function() {
...
     });
+    $('#tui-image-editor .tui-image-editor').on('mousewheel', (e) => {
+      var wDelta = e.originalEvent.wheelDelta;
+      var imageEditorWindow = e.currentTarget;
+      var initWidth = imageEditorWindow.style.width;
+      var initHeight = imageEditorWindow.style.height;
+      var newWidth;
+      var newHeight;
+      if (wDelta > 0) {
+        newWidth = parseInt(initWidth, 10) * 1.1 + 'px';
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/assets/javascripts/sitewide/file_preview.js]: Can you fix this?

File: app/assets/javascripts/sitewide/file_preview.js [hunk 2]
```diff
@@ -278,6 +276,37 @@ var FilePreviewModal = (function() {
...
+        newHeight = parseInt(initHeight, 10) * 1.1 + 'px';
+      } else {
+        newWidth = parseInt(initWidth, 10) * 0.9 + 'px';
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/assets/javascripts/sitewide/file_preview.js]: And here

File: app/assets/javascripts/sitewide/file_preview.js [hunk 3]
```diff
@@ -278,6 +276,37 @@ var FilePreviewModal = (function() {
...
+        newHeight = parseInt(initHeight, 10) * 0.9 + 'px';
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/assets/javascripts/sitewide/file_preview.js]: And here

**Comment by okriuchykhin** [comment]
Can we use bootstrap predefined screen sizes for setting image size in the editor?

**Comment by okriuchykhin** [comment]
You can add column with compress ratio to assets and populate it in post_process_file method

File: app/controllers/assets_controller.rb [hunk 4]
```diff
@@ -61,6 +61,7 @@ def file_preview
                end
 
     if @asset.is_image?
+      response_json['quality'] = @asset.quality || 90 if @asset.file.content_type == 'image/jpeg'
```
Comment (okriuchykhin) [inline app/controllers/assets_controller.rb]: I think it can be also image/pjpeg

File: app/assets/javascripts/sitewide/file_preview.js [hunk 5]
```diff
@@ -288,17 +317,21 @@ var FilePreviewModal = (function() {
...
       var bytePosition;
 
       ev.preventDefault();
       ev.stopPropagation();
 
-      imageFormat = (data['mime-type'] === 'image/png') ? 'png' : 'jpeg';
+      if (data['mime-type'] === 'image/png') {
+        imageParams = { format: 'png' };
+      } else {
+        imageParams = { format: 'jpeg', quality: (data.quality / 100) };
```
Comment (okriuchykhin) [inline app/assets/javascripts/sitewide/file_preview.js]: What if imagemagick returns quality value in different format, lets say in percents?

File: app/assets/javascripts/sitewide/file_preview.js [hunk 6]
```diff
@@ -278,6 +276,37 @@ var FilePreviewModal = (function() {
     imageEditor.on('image_loaded', () => {
       $('.file-save-link').css('display', '');
     });
+    $('#tui-image-editor .tui-image-editor').on('mousewheel', (e) => {
```
Comment (okriuchykhin) [inline app/assets/javascripts/sitewide/file_preview.js]: On initial load, how does it set image dimensions?

File: app/models/asset.rb [hunk 7]
```diff
@@ -230,6 +230,23 @@ def post_process_file(team = nil)
...
     end
+
+    if is_image?
+      Asset.delay(queue: :assets).calculate_quality(id) if file.content_type == 'image/jpeg'
+    end
+  end
+
+  def self.calculate_quality(id)
+    asset = find_by_id(id)
+    return unless asset.present? && asset.file.present?
```
Comment (okriuchykhin) [inline app/models/asset.rb]: - file? and present? checks if the _file_name field is populated
 - exists? checks if the file exists (will perform a TCP connection if stored in the cloud)

File: db/migrate/20190410110605_add_column_quality_to_asset.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddColumnQualityToAsset < ActiveRecord::Migration[5.1]
+  def change
+    add_column :assets, :quality, :integer
```
Comment (okriuchykhin) [inline db/migrate/20190410110605_add_column_quality_to_asset.rb]: I would call this column file_image_quality to be consistent with rest of the fields

File: app/models/asset.rb [hunk 9]
```diff
@@ -230,6 +230,23 @@ def post_process_file(team = nil)
       # Update asset's estimated size immediately
       update_estimated_size(team)
     end
+
+    if is_image?
+      Asset.delay(queue: :assets).calculate_quality(id) if file.content_type == 'image/jpeg'
```
Comment (okriuchykhin) [inline app/models/asset.rb]: I would try to implement it as custom processor https://github.com/thoughtbot/paperclip#custom-attachment-processors

File: app/assets/javascripts/sitewide/file_preview.js (hunk 5 above)
Comment (aignatov-bio) [reply app/assets/javascripts/sitewide/file_preview.js]: For jpeg it use quality format from 1 to 100.

Comment (okriuchykhin) [reply app/assets/javascripts/sitewide/file_preview.js]: And if in new version it moves to 0.00 - 0.99 format?

File: app/assets/javascripts/sitewide/file_preview.js (hunk 6 above)
Comment (aignatov-bio) [reply app/assets/javascripts/sitewide/file_preview.js]: By default TUI ImageEditor function:

`    resizeEditor({uiSize, imageSize = this.imageSize} = {}) {
        if (imageSize !== this.imageSize) {
            this.imageSize = imageSize;
        }
        if (uiSize) {
            this._setUiSize(uiSize);
        }

        const {width, height} = this._getEditorDimension();
        const editorElementStyle = this._editorElement.style;
        const {menuBarPosition} = this.options;

        editorElementStyle.height = `${height}px`;
        editorElementStyle.width = `${width}px`;

        this._setEditorPosition(menuBarPosition);

        this._editorElementWrap.style.bottom = `0px`;
        this._editorElementWrap.style.top = `0px`;
        this._editorElementWrap.style.left = `0px`;
        this._editorElementWrap.style.width = `100%`;

        const selectElementClassList = this._selectedElement.classList;

        if (menuBarPosition === 'top' && this._selectedElement.offsetWidth < BI_EXPRESSION_MINSIZE_WHEN_TOP_POSITION) {
            selectElementClassList.add('tui-image-editor-top-optimization');
        } else {
            selectElementClassList.remove('tui-image-editor-top-optimization');
        }
    }`

File: app/assets/javascripts/sitewide/file_preview.js (hunk 5 above)
Comment (aignatov-bio) [reply app/assets/javascripts/sitewide/file_preview.js]: Add check for format before saving

File: lib/paperclip/image_quality_calculate.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+module Paperclip
+  class ImageQualityCalculate < Processor
+    def initialize(file, options = {}, attachment = nil)
+      super
+    end
+
+    def make
+      if @file && (['image/jpeg', 'image/pjpeg'].include? @file.content_type)
+        quality = Paperclip::Processor.new(@file).identify(" -verbose #{@file.path} | grep Quality").split(':')
```
Comment (okriuchykhin) [inline lib/paperclip/image_quality_calculate.rb]: You can use "identify -format '%Q'"

File: lib/paperclip/image_quality_calculate.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        quality = (quality[1] || 90).to_f
```
Comment (okriuchykhin) [inline lib/paperclip/image_quality_calculate.rb]: I think it is always integer

**Pull request merged**